test/e2e/network/service.go

```diff
@@ -3733,84 +3695,77 @@ func validateEndpointsPorts(c clientset.Interface, namespace, serviceName string
 		if err != nil {
-			return fmt.Errorf("failed to get apiserver's restart count: %v", err)
-		}
-		if err := sshRestartMaster(); err != nil {
-			return fmt.Errorf("failed to restart apiserver: %v", err)
+			return err
 		}
-		return waitForApiserverRestarted(cs, initialRestartCount)
+		return framework.MasterUpgradeGKE(namespace, v.GitVersion[1:]) // strip leading 'v'
 	}
-	// GKE doesn't allow ssh access, so use a same-version master
-	// upgrade to teardown/recreate master.
-	v, err := cs.Discovery().ServerVersion()
-	if err != nil {
-		return err
-	}
-	return framework.MasterUpgradeGKE(namespace, v.GitVersion[1:]) // strip leading 'v'
+
+	return restartComponent(cs, kubeAPIServerLabelName, metav1.NamespaceSystem, map[string]string{clusterComponentKey: kubeAPIServerLabelName})
 }
 
-func sshRestartMaster() error {
-	if !framework.ProviderIs("gce", "aws") {
-		return fmt.Errorf("unsupported provider for sshRestartMaster: %s", framework.TestContext.Provider)
+// restartComponent restarts component static pod
+func restartComponent(cs clientset.Interface, cName, ns string, labelSet labels.Set) error {
+	pods, err := getComponentPod(cs, ns, labelSet)
+	if err != nil {
+		return fmt.Errorf("failed to get %s's pods, err: %v", cName, err)
 	}
-	var command string
-	if framework.ProviderIs("gce") {
-		command = "pidof kube-apiserver | xargs sudo kill"
-	} else {
-		command = "sudo /etc/init.d/kube-apiserver restart"
+	if err := killComponentPod(cs, ns, pods); err != nil {
+		return fmt.Errorf("failed to restart component: %s, err: %v", cName, err)
 	}
-	framework.Logf("Restarting master via ssh, running: %v", command)
-	result, err := e2essh.SSH(command, net.JoinHostPort(framework.APIAddress(), e2essh.SSHPort), framework.TestContext.Provider)
-	if err != nil || result.Code != 0 {
-		e2essh.LogResult(result)
-		return fmt.Errorf("couldn't restart apiserver: %v", err)
+	return waitForComponentRestarted(cs, cName, ns, labelSet, len(pods))
+}
+
+// killComponentPod will kill the pods
+func killComponentPod(cs clientset.Interface, ns string, pods []v1.Pod) error {
+	framework.Logf("restart pods:%v", pods)
+	for _, pod := range pods {
+		if err := wait.PollImmediate(5*time.Second, time.Minute, func() (done bool, err error) {
+			if err := cs.CoreV1().Pods(ns).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{}); err != nil {
+				framework.Logf("failed to delete pod:%v, err:%v", pod.Name, err)
+				return false, err
+			}
+			return true, nil
+		}); err != nil {
+			return fmt.Errorf("couldn't restart pod:%v, err: %v", pod.Name, err)
+		}
 	}
 	return nil
 }
 
-// waitForApiserverRestarted waits until apiserver's restart count increased.
-func waitForApiserverRestarted(c clientset.Interface, initialRestartCount int32) error {
+// waitForComponentRestarted waits until component's pod restart count increased.
+func waitForComponentRestarted(c clientset.Interface, cName, ns string, labelSet labels.Set, initialRestartCount int) error {
```

Comment:
I would get rid of this, I think it can be replaced with a call to e2epod (framework/pod) `PodsCreatedByLabel`